### ISSUE TITLE
Rule proposal: no-inline-functions-inside-jsx-expression-container

### ISSUE THREAD

**Issue by sag1v** [yannickcr/eslint-plugin-react#1761]
When inline function are created inside JSX containers (i.e `onClick={e => 'some thing'}`) the rule should warn or error about it.

The goal is to prevent "event-handlers", "render-props" or "children as a function" from creating a new function instance on each render.  

```
// bad
<div onClick={function(){}}></div>
<div onClick={function(){}()}></div>
<div onClick={() => 1}></div>
<MyComponent render={this.doSomething()} />
<MyComponent>
    {data => <MyOtherComponent data={data}></MyOtherComponent>}
</MyComponent>

// good
<div render={this.doSomething}></div>
```

possible implementation:
```
export default function(context) {
  const invalidExp = ['ArrowFunctionExpression','FunctionExpression','CallExpression'];
  return {
    JSXExpressionContainer({expression}){
      if(invalidExp.includes(expression.type)){
        context.report(expression, 'Do not use inline functions inside JSX Expression Container');
      }
    }
  };
}
```

I'm no expert with writing AST's but as far as i know the type `JSXExpressionContainer` will only be used inside `render` calls context.

**Comment by sag1v** [comment]
Fair enough, I didn't notice that one.

**Comment by sbrighiu** [comment]
jsx-no-bind shows false positives when a component is declared inside the Component and just referenced like 'onPress={this.onPress} ...'

I would also want to make this a linting run inside my project... the only requirement is no inline functions declared in jsx. They should be able to be declared in the same Component (react - react native)

**Comment by ljharb** [comment]
@sbrighiu please file an issue for the false positives.

**Comment by sbrighiu** [comment]
https://github.com/yannickcr/eslint-plugin-react/issues/1736 I think this is a bit similar to what I am experiencing. Since i am using react hooks, i need to declare the function and save its reference to use in jsx code. This is by now common practic, and i would have expected the rule to just inform me when i declare the function inside the jsx code, inside the Component props. 

Please correct me if i'm wrong

**Comment by sbrighiu** [comment]
This is why i believe @sag1v was onto something. This rule could also benefit https://github.com/yannickcr/eslint-plugin-react/issues/629, non-jsx code, enforcing that inline functions should not be allowed.

**Comment by ljharb** [comment]
That's what the `useCallback` hook is for.

**Comment by sbrighiu** [comment]
well yes but useCallback only works correctly when also using React.memo. Not every function needs to be memoized, some just need their reference saved.

**Comment by ljharb** [comment]
I'm reasonably sure that's not correct. `useCallback` is just a special form of `useMemo`, and does not require React.memo to be used at all.

**Comment by ljharb** [comment]
using useCallback ensures that if the components you pass those callbacks *to* use useMemo, there’s an improvement. Thus, it’s always a best practice to use it, every time, just in case.

**Comment by ljharb** [comment]
if you don’t have perf issues, then this lint rule wouldn’t be particularly useful either, and lint rules generally are only useful when they always or never apply.
